### PR TITLE
Throw error if setting static grads to `None` in `zero_grad()`

### PR DESCRIPTION
**Pull request by mlazos**
If a tensor is marked a static address for dynamo, this implies that the pointer will be the same across calls to the optimizer step. For this constraint to hold, we should not set grads marked as static to None. 

After some discussion, I opted to throw an error.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @Xia-Weiwen

### CODE REVIEW

**Comment by mlazos** [comment]
I'm open to making this ternary - ie it's None by default and leaves static tensors alive, but if the user explicitly sets True then set even static tensors to None.

**Comment by ezyang** [comment]
Let's talk about this some more.

I think, in principle, the general idea here is OK, which is that a more CUDA graph friendly optimizer interface would avoid zeroing out gradients so that you can use the same memory address for gradients across runs.

However, there are two problems:

1. I don't actually need to hold on to the memory here for the entire training loop. Even in CUDA graphs, I need to "deallocate" the gradient space so that I can recycle this memory for other intermediate activations before I actually populate grad. So even if you have a problem, this seems like a questionable way to implement it.
2. Even if a user wants to keep the gradient tensor live across iterations, they should explicitly ask for it, it shouldn't automatically happen due to some internal Dynamo config. It seems like we already have this knob.

Can you tell us more about what problem you're trying to solve? You also don't have any tests, nor a referenced issue in the PR.

**Comment by mlazos** [comment]
> Let's talk about this some more.
> 
> I think, in principle, the general idea here is OK, which is that a more CUDA graph friendly optimizer interface would avoid zeroing out gradients so that you can use the same memory address for gradients across runs.
> 
> However, there are two problems:
> 
> 1. I don't actually need to hold on to the memory here for the entire training loop. Even in CUDA graphs, I need to "deallocate" the gradient space so that I can recycle this memory for other intermediate activations before I actually populate grad. So even if you have a problem, this seems like a questionable way to implement it.
> 2. Even if a user wants to keep the gradient tensor live across iterations, they should explicitly ask for it, it shouldn't automatically happen due to some internal Dynamo config. It seems like we already have this knob.
> 
> Can you tell us more about what problem you're trying to solve? You also don't have any tests, nor a referenced issue in the PR.

Yeah I meant for this to start a discussion, I'll def add tests.

The problem I'm trying to solve is if the upstream model is not cudagraphed (say someone wants to run the optimizer standalone and see the perf). There currently was not a mechanism before the static address API to prevent additional copies that cudagraphs introduces if it can't guarantee that the addresses are static. My next thought after introducing this was that zero_grad deallocating tensors would induce constant recompiles if it were to reallocate the grads on each iteration (since dynamo will guard on the addresses of static tensors)

**Comment by janeyx99** [comment]
I agree with Ed's points that grads today are more like temporary intermediates during training (vs persisting across loops) and that we should consider a more cudagraph friendly API for users to opt into.

The current PR would change the behavior for users in a unexpected way through an internal dynamo knob, which I also agree is not the right approach. We should have a way for users to explicitly tick when they want this behavior (and have the memory to sustain it) and in the default case we should follow the current semantic.

**Comment by mlazos** [comment]
> I agree with Ed's points that grads today are more like temporary intermediates during training (vs persisting across loops) and that we should consider a more cudagraph friendly API for users to opt into.
> 
> The current PR would change the behavior for users in a unexpected way through an internal dynamo knob, which I also agree is not the right approach. We should have a way for users to explicitly tick when they want this behavior (and have the memory to sustain it) and in the default case we should follow the current semantic.

This is that knob, the user explicitly needs to call `mark_static_address(t)` on a tensor to mark it as static if the memory is available. If they don't use that API, the grads will get deallocated as usual, dynamo will not call it automatically.

**Comment by janeyx99** [comment]
> There currently was not a mechanism before the static address API to prevent additional copies of every grad that cudagraphs introduces if it can't guarantee that the addresses are static, which is catastrophic for perf. 

Does this mean that if people do not use `mark_static_address(p.grad)`, using CUDA graphs would be not performant?

Regarding the use case, would this be for compiling the full training loop? Since the grads are expected to exist for the optimizer.step() and people don't normally run optimizer.step() multiple times in a row, would this API be more relevant for CUDA-graphing the whole training loop? If so, I'm curious how running CUDA graphs on the whole training loop today in eager treat grads. In the compiler world, I would expect grad to be introduced as an intermediate versus something that can cause a recompile. 

Another use case is perhaps if someone CUDA graphs only the optimizer step and not anything before. In that case, the grad is treated as an intermediate before optimizer.step() and will be different addresses every loop. Even in this case, wouldn't the CUDA graph persist for just the optimizer step part and take as input the new grad addresses each time? Within the step, the addresses wouldn't change, right?

I'd like to understand the use case for `mark_static_address(p.grad)` better. Implementation-wise, I'd rather the code error with a warning telling the user to use optimizer.zero_grad(set_to_None=False) instead of disabling that behavior under the hood.

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -805,6 +806,10 @@ def zero_grad(self, set_to_none: bool = True) -> None:
                 for p in group['params']:
                     if p.grad is not None:
                         if set_to_none:
+                            if get_static_address_type(p.grad) is not None:
+                                raise RuntimeError("zero_grad(set_to_none=True) only expects grads that"
+                                                   "are not marked as tensors with static addresses. "
+                                                   "Use zero_grad(set_to_none=False) instead.")
```
Comment (janeyx99) [inline torch/optim/optimizer.py:812]:
```suggestion
                                raise RuntimeError("Default zero_grad(set_to_none=True) will not respect grads that"
                                                   "are marked as tensors with static addresses in dynamo and will "
                                                   "set them to None. Use zero_grad(set_to_none=False) instead.")
```

**Comment by mlazos** [comment]
@janeyx99 and I discussed offline and decided the best approach was throw an error if zero_grad(set_to_none=True) is called with static tensors.

File: torch/optim/optimizer.py [hunk 2]
```diff
@@ -789,6 +789,7 @@ def zero_grad(self, set_to_none: bool = True) -> None:
                 (in one case it does the step with a gradient of 0 and in the other it skips
                 the step altogether).
         """
+        from torch._dynamo.utils import get_static_address_type
```
Comment (janeyx99) [inline torch/optim/optimizer.py:792]: Should this be within the zero_grad function or outside?

Also, there are other places people can call zero_grad (e.g., module.zero_grad). Just FYI.

**Review by janeyx99 — COMMENTED** [review]
Could you add a test case in either test_cuda or test_optim or wherever we should test the mark_static_address API?

File: torch/optim/optimizer.py (hunk 2 above)
Comment (mlazos) [reply torch/optim/optimizer.py:792]: There is a circular dependency issue with importing torch._dynamo.utils in the optimizer module, this is why all the dynamo imports in this file are local. It's actually non-trivial to resolve as well so we usually duplicate APIs to torch.utils. This api seemed pretty dynamo specific, and I'm only using this in one place so I didn't think it was worth it in this case.

File: test/dynamo/test_decorators.py [hunk 3]
```diff
@@ -303,6 +303,18 @@ def test_mark_static_address_guarded(self):
...
+    def test_optimizer_zero_grad_static_err(self):
+        mod = torch.nn.Linear(10, 10)
+
+        for param in mod.parameters():
+            param.grad = torch.randn_like(param)
+            torch._dynamo.mark_static_address(param.grad)
+
+        optim = torch.optim.SGD(mod.parameters(), lr=1e-2)
+
+        with self.assertRaises(RuntimeError):
```
Comment (janeyx99) [inline test/dynamo/test_decorators.py:315]: could we do an assertRaisesRegex here so people know what this is testing

File: test/dynamo/test_decorators.py [hunk 4]
```diff
@@ -303,6 +303,18 @@ def test_mark_static_address_guarded(self):
...
+            optim.zero_grad(set_to_none=True)
```
Comment (janeyx99) [inline test/dynamo/test_decorators.py:316]:
```suggestion
            optim.zero_grad()
```

should be default

File: torch/optim/optimizer.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/optimizer.py:792]: Is it in any roadmap to resolve the circular dependency?

Comment (albanD) [inline torch/optim/optimizer.py:792]: `_dynamo` is a lazy import no? This will actually trigger the lazy import for users that don't use dynamo. that sounds bad?

Comment (albanD) [reply torch/optim/optimizer.py:792]: You could just use `torch._dynamo.utils.get_static_address_type()` directly below to remove this import

**Review by albanD — COMMENTED** [review]
Also note that these is another zero_grad on Module.

File: torch/optim/optimizer.py (hunk 2 above)
Comment (mlazos) [reply torch/optim/optimizer.py:792]: Good catch, I think I'll move this to torch._utils instead
